datafusion/core/src/execution/context.rs

```diff
@@ -613,50 +615,29 @@ impl SessionContext {
     /// [`read_table`](Self::read_table) with a [`ListingTable`].
     async fn _read_type<'a>(
         &self,
-        table_path: impl AsRef<str>,
+        table_paths: Vec<impl AsRef<str>>,
         options: impl ReadOptions<'a>,
     ) -> Result<DataFrame> {
-        let table_path = ListingTableUrl::parse(table_path)?;
+        let table_paths = table_paths
+            .iter()
+            .map(ListingTableUrl::parse)
+            .collect::<Result<Vec<ListingTableUrl>>>()?;
         let session_config = self.copied_config();
         let listing_options = options.to_listing_options(&session_config);
         let resolved_schema = match options
-            .get_resolved_schema(&session_config, self.state(), table_path.clone())
+            .get_resolved_schema(&session_config, self.state(), table_paths[0].clone())
             .await
         {
             Ok(resolved_schema) => resolved_schema,
             Err(e) => return Err(e),
         };
-        let config = ListingTableConfig::new(table_path)
+        let config = ListingTableConfig::new_with_multi_paths(table_paths)
             .with_listing_options(listing_options)
             .with_schema(resolved_schema);
         let provider = ListingTable::try_new(config)?;
         self.read_table(Arc::new(provider))
     }
 
-    /// Creates a [`DataFrame`] for reading an Avro data source.
-    ///
-    /// For more control such as reading multiple files, you can use
-    /// [`read_table`](Self::read_table) with a [`ListingTable`].
-    pub async fn read_avro(
-        &self,
-        table_path: impl AsRef<str>,
-        options: AvroReadOptions<'_>,
-    ) -> Result<DataFrame> {
-        self._read_type(table_path, options).await
-    }
```

Comment:
I wonder if you could leave these methods on `SessionContext` and instead use a trait for the argument and avoid changes to downstream users. 

Perhaps something like:


```rust
trait DataFilePaths {
  // Parse to a list of URs
  fn to_urls(self) -> Result<Vec<ListingTableUrl>>>;
}

impl DataFilePaths for &str {
  fn to_urls(self) -> Result<Vec<ListingTableUrl>>> {
    Ok(vec![ListingTableUrl::parse(self)?])
  }
}

impl DataFilePaths for Vec<&str> {
  fn to_urls(self) -> Result<Vec<ListingTableUrl>>> {
    self
            .iter()
            .map(ListingTableUrl::parse)
            .collect::<Result<Vec<ListingTableUrl>>>()
  }
}



impl SessionContext {
...
    pub async fn <P> read_avro(
        &self,
        table_path: P,
        options: AvroReadOptions<'_>,
    ) -> Result<DataFrame> {
        self._read_type(table_path.to_urls()>, options).await
    }
...
```

Reply:
This might be better. Less code to implement. But we have to use the `use` statement for the defined trait whenever we call `read_csv` and others.

One more issue could be when the users implement a trait with the same method name. They won't be able to use both traits, only one.

Reply:
Hi @alamb, I was travelling and couldn't work on this last week. In the above example, we will need to use the `use` statement even when we are using `read_csv("s3://bucket/key", options)`. Instead I propose the following:
1. What if we implement AsRef for Vec. Then, we only need to use the `use` statement when we are calling for multiple  paths. The default single path code will remain unchanged. The AsRef will simply concatenate all the strings in Vec with `,` delimiter. And in the `_read_type`, we will split on `,`.
2. How about having a different method `read_csvs` for taking multiple paths. `read_csvs` is a lot cleaner than `read_csv_with_multi_paths`.

Reply:
> How about having a different method read_csvs for taking multiple paths. read_csvs is a lot cleaner than read_csv_with_multi_paths.

I agree that would be better

I think my concern was about the fact that then there would be `read_parquets` `read_avros`, etc. which starts to sound like too many 🤔

Reply:
Looks like a toss up between multiple methods and using traits. Shall we get inputs from other core members?

Reply:
Yes, I think that would be a good idea. Perhaps @andygrove @Dandandan @yahoNanJing @liukun4515  or @tustvold  have some thoughts

Reply:
> But we have to use the use statement for the DataFilePaths whenever we call read_csv and others.

I'm surprised by this, you shouldn't need to use a trait in order to satisfy a type constraint on a generic method?

Reply:
The generic method would have a constraint on P as `P: DataFilePaths`. When we call the method `read_csv("path", options)` we will have to use the `use ..DataFilePaths` so that rust knows the trait `DataFilePaths` is implemented for string. Is this what you meant?

Reply:
That isn't how traits behave, you do not need to have a trait in scope in order to call a generic method with a type constraint on it - https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a629bf1bdd4b6418ab09e274f8214de5

Reply:
Thanks. I didn't know traits worked that way with generics. I tried an implementation and it worked!